### PR TITLE
ENH: Merge empty hist bins

### PR DESCRIPTION
**Pull request by areeves87**
ENH/BUG: merge empty bins when np.histogram(bin='auto') to avoid raising MemoryErrors

See #11879 , #10297, #8203 

histogram(bins='auto') will sometimes raise a memory error if the number of automatically-generated bin edges is too large. In all documented cases, the conditions producing outsized bin numbers are when the auto-binning defaults to the 'fd' method.

Based on suggestions from @eric-wieser , this approach merges empty bins. In order to minimize the impact of this enhancement, I only invoke the new method when 'auto' defaults to 'fd'. A direct call to the 'fd' method produces the same results as before.

Note that this approach gives us back unequal bin widths. The non-empty bins all have the same width, but the zero bins will vary in their width depending on how many empty bins there are. 

All tests are passing. I did have to change the expected number of bins for one test to account for the merged empty bins. You can see this in `test_histograms.simple_range`. I also added a new test that expects a direct call to the 'fd' method to raise a MemoryError, and a call to bins='auto' to produce a small number of bins.

### CODE REVIEW

**Comment by mattip** [comment]
Strange, there is a failure but only on PyPy.

<details>

```
=================================== FAILURES ===================================
_________________ TestHistogramOptimBinNums.test_simple_range __________________

self = <numpy.lib.tests.test_histograms.TestHistogramOptimBinNums object at 0x0000000008cb12b8>

    def test_simple_range(self):
        """
        Straightforward testing with a mixture of linspace data (for
        consistency). Adding in a 3rd mixture that will then be
        completely ignored. All test values have been precomputed and
        the shouldn't change.
        """
        # some basic sanity checking, with some fixed data.
        # Checking for the correct number of bins
        basic_test = {
                      50:   {'fd': 8,  'scott': 8,  'rice': 15,
                             'sturges': 14, 'auto': 14, 'stone': 8},
                      500:  {'fd': 15, 'scott': 16, 'rice': 32,
                             'sturges': 20, 'auto': 20, 'stone': 80},
                      5000: {'fd': 33, 'scott': 33, 'rice': 69,
                             'sturges': 27, 'auto': 19, 'stone': 80}
                     }
    
        for testlen, expectedResults in basic_test.items():
            # create some sort of non uniform data to test with
            # (3 peak uniform mixture)
            x1 = np.linspace(-10, -1, testlen // 5 * 2)
            x2 = np.linspace(1, 10, testlen // 5 * 3)
            x3 = np.linspace(-100, -50, testlen)
            x = np.hstack((x1, x2, x3))
            for estimator, numbins in expectedResults.items():
                a, b = np.histogram(x, estimator, range = (-20, 20))
                msg = "For the {0} estimator".format(estimator)
                msg += " with datasize of {0}".format(testlen)
>               assert_equal(len(a), numbins, err_msg=msg)
E               AssertionError: 
E               Items are not equal: For the auto estimator with datasize of 5000
E                ACTUAL: 33
E                DESIRED: 19

a          = array([  0,   0,   0,   0,   0,   0,   0,   0, 202, 270, 269, 269, 269,
       270, 269, 182,   0, 273, 404, 404, 404, 404, 404, 404, 303,   0,
         0,   0,   0,   0,   0,   0,   0])
b          = array([-20.        , -18.78787879, -17.57575758, -16.36363636,
       -15.15151515, -13.93939394, -12.72727273, -11.51...15152,  12.72727273,
        13.93939394,  15.15151515,  16.36363636,  17.57575758,
        18.78787879,  20.        ])
basic_test = {50: {'auto': 14, 'fd': 8, 'rice': 15, 'scott': 8, ...}, 500: {'auto': 20, 'fd': 15, 'rice': 32, 'scott': 16, ...}, 5000: {'auto': 19, 'fd': 33, 'rice': 69, 'scott': 33, ...}}
estimator  = 'auto'
expectedResults = {'auto': 19, 'fd': 33, 'rice': 69, 'scott': 33, ...}
msg        = 'For the auto estimator with datasize of 5000'
numbins    = 19
self       = <numpy.lib.tests.test_histograms.TestHistogramOptimBinNums object at 0x0000000008cb12b8>
testlen    = 5000
x          = array([-10.        ,  -9.99549775,  -9.9909955 , ..., -50.020004  ,
       -50.010002  , -50.        ])
x1         = array([-10.        ,  -9.99549775,  -9.9909955 , ...,  -1.0090045 ,
        -1.00450225,  -1.        ])
x2         = array([ 1.      ,  1.003001,  1.006002, ...,  9.993998,  9.996999,
       10.      ])
x3         = array([-100.      ,  -99.989998,  -99.979996, ...,  -50.020004,
        -50.010002,  -50.      ])

```

**Comment by areeves87** [comment]
Do I need to do something on my end? All tests pass when I run 
```
python3 runtests.py -v -t numpy/lib/tests/test_histograms.py
```
On my machine.

File: numpy/lib/histograms.py [hunk 1]
```diff
@@ -436,6 +505,10 @@ def _get_bin_edges(a, bins, range, weights):
             bin_type = np.result_type(bin_type, float)
 
         # bin edges must be computed
+        if all([bins is 'auto', auto_method is 'fd', n_equal_bins>2]):
```
Comment (mattip) [inline numpy/lib/histograms.py]: Test for string equality with `==` : `bins == 'auto', auto_method == 'fd'`

Comment (seberg) [reply numpy/lib/histograms.py]: Also  use `and` and not `all`, not just easier to read, also faster.

File: numpy/lib/histograms.py [hunk 2]
```diff
@@ -262,6 +262,36 @@ def _hist_bin_auto(x, range):
...
+    -------
+    h : An estimate of the optimal bin width for the given data.
+
+    See Also
+    --------
+    _hist_bin_fd, _hist_bin_sturges
+    """
+    fd_bw = _hist_bin_fd(x, range)
+    sturges_bw = _hist_bin_sturges(x, range)
+    del range  # unused
```
Comment (mattip) [inline numpy/lib/histograms.py]: No need for this line

File: numpy/lib/histograms.py [hunk 3]
```diff
@@ -262,6 +262,36 @@ def _hist_bin_auto(x, range):
...
+
+    Parameters
+    ----------
+    x : array_like
+        Input data that is to be histogrammed, trimmed to range. May not
+        be empty.
+
+    Returns
+    -------
+    h : An estimate of the optimal bin width for the given data.
```
Comment (mattip) [inline numpy/lib/histograms.py]: not accurate - this returns either 'fd' or 'sturges' - whichever has the smaller number of bins

File: numpy/lib/histograms.py [hunk 4]
```diff
@@ -349,6 +379,41 @@ def _unsigned_subtract(a, b):
...
+    """
+    keep = (a >= first_edge)
+    keep &= (a <= last_edge)
+    if not np.logical_and.reduce(keep):
+        a = a[keep]
+
+    if a.size == 0:
+        return np.array([0,1])
+
+    width = np.ptp((first_edge,last_edge))/(n_equal_bins)
```
Comment (mattip) [inline numpy/lib/histograms.py]: can `n_equal_bins` ever be 0?

File: numpy/lib/histograms.py [hunk 5]
```diff
@@ -349,6 +379,41 @@ def _unsigned_subtract(a, b):
...
+    left_edges = np.floor_divide((a - first_edge), width) * width
```
Comment (mattip) [inline numpy/lib/histograms.py]: can width ever be 0?

File: numpy/lib/histograms.py (hunk 4 above)
Comment (mattip) [reply numpy/lib/histograms.py]: ahh, I see not, since this is only called when `n_equal_bins > 2`.  Maybe add that to the parameter description above

File: numpy/lib/histograms.py (hunk 5 above)
Comment (areeves87) [reply numpy/lib/histograms.py]: For the width calculation, `first_edge` and `last_edge` are forwarded arguments from `_get_outer_edges`. There are several checks in that function that would prevent `np.ptp((first_edge,last_edge))` from being 0. For example, if `first_edge` and `last_edge` are found to be equal in `_get_outer_edges`, then just before it returns their values, it performs this expansion:

```
# expand empty range to avoid divide by zero
    if first_edge == last_edge:
        first_edge = first_edge - 0.5
        last_edge = last_edge + 0.5
```

I think it is safe to say that width cannot be zero here.

**Comment by mattip** [comment]
LGTM. Does anyone else want to take a look? This closes the three issues at the top of the PR

File: numpy/lib/histograms.py [hunk 6]
```diff
@@ -349,6 +378,42 @@ def _unsigned_subtract(a, b):
...
+    right_edges = left_edges + width
+    inner_edges = np.unique((left_edges, right_edges)) + first_edge
+    edges = np.append(np.append(first_edge, inner_edges), last_edge)
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Does this essentially compute the histogram twice now?

Comment (areeves87) [reply numpy/lib/histograms.py]: Can you elaborate? I'm not sure what you mean.

Maybe it seems that way because there are steps for both left AND right bin edges? These two steps create a left and right bin edge for each data point. The left edges are computed from the data but the right edges are computed from the left edges. So I wouldn't say the histogram edges get computed twice. Instead, I would say the left edges are the only ones computed from the data.

Am I addressing the right question?

Another possibility is that you are concerned this new method re-calculates bins that were already computed in `_get_bin_edges.` That is also not the case. The new method is only reached when `bins=='auto'` defaults to `'fd'`, and returns the bin edges before the other bin edge calculation is reached.

Maybe my code is hard to follow. As an example of how the right and left bin edge calculation works, consider this step by step process:

- width = 2 # pre-calculated bin width
- [0,1,2,9] # original data
- [0,0,1,4] # floor division by bin width
- [0,0,2,8] # "left edges" of histogram bins, computed by multiplying floored data by bin width (I suppose you could call this the "quantization" step ala signal processing.)
- [0,2,8]   # unique left edges of histogram bins
- [2,4,10]  # "right edges" of histogram bins (left edges + bin width)
- [0,2,4,8,10] # combined edges of histogram bins, with the empty bins (4,6) and (6,8) merged into an empty bin spanning (4,8)

That is the gist of the method I propose. I add some extra steps for edge cases.

File: numpy/lib/histograms.py [hunk 7]
```diff
@@ -349,6 +378,44 @@ def _unsigned_subtract(a, b):
...
+    """
+    keep = (a >= first_edge)
+    keep &= (a <= last_edge)
+    if not np.logical_and.reduce(keep):
+        a = a[keep]
+
+    if a.size == 0:
+        return np.array([0,1])
+
+    width = np.ptp((first_edge,last_edge))/(n_equal_bins)
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: `ptp` is overkill here for subtracting two scalars, one of which you know to be larger than the other.
```suggestion
    width = _unsigned_subtract(last_edge, first_edge) / n_equal_bins
```

File: numpy/lib/histograms.py [hunk 8]
```diff
@@ -349,6 +378,44 @@ def _unsigned_subtract(a, b):
...
+    zeroed_min = a - first_edge
+    quantized = np.floor_divide(zeroed_min, width) * width
+    left_edges = np.unique(quantized)
+    right_edges = left_edges + width
+    unzeroed_edges = np.unique((left_edges, right_edges)) + first_edge
+    edges = np.concatenate(([first_edge], unzeroed_edges, [last_edge]))
+
+    TOL = width/2
+    widths = np.append(width, np.diff(edges)) #remove bins produced by float errors
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: `append` is needless indirection here
```suggestion
    widths = np.concatenate(([width], np.diff(edges)))  # remove bins produced by float errors
```

File: numpy/lib/histograms.py [hunk 9]
```diff
@@ -349,6 +378,44 @@ def _unsigned_subtract(a, b):
...
+    if a.size == 0:
+        return np.array([0,1])
+
+    width = np.ptp((first_edge,last_edge))/(n_equal_bins)
+    zeroed_min = a - first_edge
+    quantized = np.floor_divide(zeroed_min, width) * width
+    left_edges = np.unique(quantized)
+    right_edges = left_edges + width
+    unzeroed_edges = np.unique((left_edges, right_edges)) + first_edge
+    edges = np.concatenate(([first_edge], unzeroed_edges, [last_edge]))
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: I'm rather worried that there's floating point rounding that can happen here which will push data into the "empty" bin that you combined.

File: numpy/lib/histograms.py [hunk 10]
```diff
@@ -349,6 +378,44 @@ def _unsigned_subtract(a, b):
...
+    if not np.logical_and.reduce(keep):
+        a = a[keep]
+
+    if a.size == 0:
+        return np.array([0,1])
+
+    width = _unsigned_subtract(last_edge, first_edge) / n_equal_bins
+    zeroed_min = a - first_edge
+    quantized = np.floor_divide(zeroed_min, width) * width
+    left_edges = np.unique(quantized)
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Regarding my remark about computing the histogram twice - I think adding `return_counts=True` would do that for you here, at a fairly low cost.

Comment (eric-wieser) [reply numpy/lib/histograms.py]: (not to say necessarily that it's a good idea)

File: numpy/lib/histograms.py (hunk 9 above)
Comment (areeves87) [reply numpy/lib/histograms.py]: I'm adding an extra set of edges to avoid push data into empty bins now. A bit inelegant but it works in the cases I tested.

**Comment by seberg** [comment]
Can you propose this on the mailing list? It sounds like the main issue is the decision to extend the API in this way. One possible reason against it: If you use `"auto"` once, and then re-use the bins, the uneven bins may not be what you want. Since the strings are a utility feature, so that is likely OK. I personally do not have a big opinion about this.

**Comment by eric-wieser** [comment]
For the record, my stance is that idea is sound, but I think the floating-point devil is in the details, so we need to review this very carefully to avoid reintroducing a regression.

Perhaps `hypothesis` can help find the corner-cases.

**Comment by areeves87** [comment]
Okay I sent a message to the discussion group here https://mail.python.org/pipermail/numpy-discussion/2020-February/080388.html

**Comment by tacaswell** [comment]
>  One possible reason against it: If you use "auto" once, and then re-use the bins, the uneven bins may not be what you want. 

This is something that Matplotlib does in `ax.hist` if the user passes in multiple datasets.  We are fine with the un-even bins, but the bins that work for the first one may be very wrong for the rest.

**Comment by tacaswell** [comment]
Sorry, posted my last message as the train was pulling into the station and was too hasty.  We compute the bins based on the concatenated data sets, not just the first one (at least on master).

**Comment by mattip** [comment]
We discussed this at a triage meeting, and reached the conclusion that a better approach would be to properly allocate a reasonable number of bins at the outset and not merge bins later on. We now automatically limit the number of bins. Closing.
